### PR TITLE
feat(sql): support multiple columns in upsert's `on` parameter

### LINKED ISSUE #11996: feat: multiple 'on' columns in upsert
### Is your feature request related to a problem?

I'm missing the ability to do upsert on multiple columns. Currently On param support a single string - i.e a single column name. 

### What is the motivation behind your request?

upserting on multiple key for example tenant_id, entity_id

### Describe the solution you'd like

support to optionally put list of Column object or strings. 

### What version of ibis are you running?

12.0.0

### What backend(s) are you using, if any?

BigQuery 
DuckDB

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bhargaviHQ**
## Description of changes
 `upsert`'s `on` parameter now accepts an iterable of column names in addition to a single string, so upserting on a composite key (e.g. `tenant_id, entity_id`) is possible. The `WHEN MATCHED` update clause excludes all key columns, and the `ON` join condition ANDs an equality per key column. Single-column `on` behavior is unchanged.

## Issues closed
* Resolves #11996

## Test plan
  - [x] Added `test_upsert_from_dataframe_multiple_on_columns`, designed so single- vs. multi-column `on` produce genuinely different results
  - [x] Added `test_upsert_empty_on_raises` for the `on=[]` edge case (raises a clear `IbisInputError` instead of an internal sqlglot `ValueError`)
  - [x] Verified both new tests fail without the fix and pass with it 
  - [x] Inspected generated `MERGE` SQL directly for both cases; confirmed single-column `on` produces unchanged SQL
  - [x] `pytest ibis/backends/tests/test_client.py -k duckdb` (195 passed)

### CODE REVIEW

File: ibis/backends/tests/test_client.py [hunk 1]
```diff
@@ -698,6 +698,46 @@ def test_upsert_from_dataframe(
...
+            "last_name": ["Z", "Y"],
+            "department_name": ["NEW1", "NEW2"],
+            "salary": [999.0, 888.0],
+        }
+    )
+
+    con.upsert(employee_data_1_temp_table, obj=source, on=["first_name", "last_name"])
+    result = temporary.execute()
+
+    expected = pd.concat([df1, source], ignore_index=True)
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py:723]: Need to test something where the merge result isn't just a union.

**Review by deepyaman — CHANGES_REQUESTED** [review]
Looks reasonable, will take a closer look. One thing need is a better test case. (Also, does the test case need to be separate? Or could parametrize it on the existing one.)

File: ibis/backends/tests/test_client.py (hunk 1 above)
Comment (bhargaviHQ) [reply ibis/backends/tests/test_client.py]: Thanks for reviewing @deepyaman. Merged both tests into one and added a case where a row matches on both columns and gets updated.

**Review by deepyaman — COMMENTED** [review]
Superseded by [my review below](https://github.com/ibis-project/ibis/pull/12123#pullrequestreview-5270547173).

<details>
<summary>Original comment</summary>

Thanks, the implementation looks correct to me. Single-column `on` generates the exact same `MERGE` as `main`, and the multi-column case ANDs the keys and excludes them from `UPDATE SET` as expected. A few things on the tests, plus one edge case:

**1. The multi-column test doesn't exercise the second key column.** If I change the `ON` clause to only join on `on_columns[:1]` (i.e. just `first_name`), both params still pass: `("B", "E")` matches on `first_name` alone anyway, and `"X"` is new either way. The earlier version had the right idea with a row that shares `first_name` but not `last_name`; that got lost when switching to an update case. Including both catches it (verified it fails with the mutation above):

```python
# update; insert despite matching first_name; insert
"first_name": ["B", "B", "X"],
"last_name": ["E", "Z", "Y"],
"department_name": ["ZZ1", "ZZ2", "ZZ3"],
"salary": [999.0, 888.0, 777.0],
```

A one-line comment like the one above is plenty; no need to explain how upsert works.

**2. `test_upsert_empty_on_raises`:** the comment isn't accurate: `upsert()` registers the memtable (`_run_pre_execute_hooks`) before `_build_upsert_from_table` validates `on`, so backend SQL does run first. I'd move the validation up into `upsert()` before the memtable registration and drop the comment; the test body is self-explanatory.

**3. Nit:** `(on,) if isinstance(on, str) else tuple(on)` can be `util.promote_tuple(on)`.

**4. Edge case (fine as a follow-up):** if every column is a key column (e.g. a junction table upserted on `(a, b)`), this generates `WHEN MATCHED THEN UPDATE` with no `SET`. DuckDB happens to accept that, but e.g. Postgres requires `UPDATE SET ...`. This was technically reachable before with a one-column table, but composite keys make it much more likely. Probably simplest to omit the `WHEN MATCHED` clause when there's nothing to update.



---

**Update**: I checked all of this against Postgres and Oracle in addition to DuckDB, and need to correct point 4. DuckDB doesn't just tolerate the bare `WHEN MATCHED THEN UPDATE`; it treats it as `UPDATE SET *` by position. So with this PR, upserting a two-column table on both columns silently swaps values if the source columns are in a different order: a matched row `(a=1, b=10)` with the source columns ordered `(b, a)` ends up as `(10, 1)`. The same bare `UPDATE` also gets generated when the source has only the key columns, e.g. source `(a, b)` into a table `(a, b, c)`. DuckDB errors on that with a column-count mismatch, Postgres gives `syntax error at or near "WHEN"`, and Oracle gives `ORA-00971: missing SET keyword`.

Since multi-column `on` makes this much easier to hit, I think it should be fixed in this PR rather than as a follow-up: skip the `WHEN MATCHED` clause when there are no non-key columns to update, and add a test for the key-only source case. I tried that change locally, and the upsert tests still pass on all three backends.

Also, a small clarification on point 1: with the suggested data and an `ON` that only joins on the first key, Postgres and Oracle fail with a "MERGE cannot update the same row twice" error rather than an assertion. The test still catches it either way.



</details>

File: ibis/backends/sql/__init__.py [hunk 2]
```diff
@@ -602,13 +602,17 @@ def _build_upsert_from_table(
...
         source,
-        on: str,
+        on: str | Iterable[str],
         db: str | None = None,
         catalog: str | None = None,
     ):
         compiler = self.compiler
         quoted = compiler.quoted
 
+        on_columns = (on,) if isinstance(on, str) else tuple(on)
```
Comment (deepyaman) [inline ibis/backends/sql/__init__.py]: Nit: this can use `util.promote_tuple`.

```suggestion
        on_columns = util.promote_tuple(on)
```

File: ibis/backends/sql/__init__.py [hunk 3]
```diff
@@ -602,13 +602,17 @@ def _build_upsert_from_table(
...
+        if not on_columns:
```
Comment (deepyaman) [inline ibis/backends/sql/__init__.py]: Can we move this check up into `upsert()`, before `self._run_pre_execute_hooks(obj)`? Right now the memtable gets created and populated in the backend (and `get_schema` would run too) before we find out `on` is empty. BigQuery's override just forwards to `super().upsert()`, so it'd be covered too.

File: ibis/backends/sql/__init__.py [hunk 4]
```diff
@@ -624,7 +628,7 @@ def _build_upsert_from_table(
                             sg.column(col, table=source_alias, quoted=quoted)
                         )
                         for col in columns
-                        if col != on
+                        if col not in on_columns
```
Comment (deepyaman) [inline ibis/backends/sql/__init__.py]: If there are no non-key columns left, this produces a bare `WHEN MATCHED THEN UPDATE` with no `SET`. You can hit that by upserting a source that only has the key columns, e.g. source `(a, b)` into a table `(a, b, c)` with `on=["a", "b"]`, or a table that's all key columns.

Postgres fails with `syntax error at or near "WHEN"` and Oracle with `ORA-00971: missing SET keyword`. DuckDB is worse, because it treats a bare `UPDATE` as a positional `UPDATE SET *`. With a table `(a, b)` and a source whose columns are ordered `(b, a)`, a matched row `(1, 10)` silently becomes `(10, 1)`.

This was technically possible before with a single-column table, but composite keys make it a lot easier to hit. I think the simplest fix is to leave out the `WHEN MATCHED` clause when there's nothing to update. I tried that locally and it works on DuckDB, Postgres, and Oracle. Could you add a test for the key-only source case too?

File: ibis/backends/tests/test_client.py [hunk 5]
```diff
@@ -681,23 +681,64 @@ def _emp(a, b, c, d):
...
+            ["first_name", "last_name"],
+            pd.DataFrame(
+                {
+                    # ("B", "E") matches an existing row on both columns and
+                    # should be updated; ("X", "Y") matches nothing and
+                    # should be inserted
+                    "first_name": ["B", "X"],
+                    "last_name": ["E", "Y"],
+                    "department_name": ["ZZ1", "ZZ2"],
+                    "salary": [999.0, 888.0],
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py]: This doesn't actually exercise the second key column. If I change the `ON` clause to only join on `first_name`, the test still passes, since `("B", "E")` matches on `first_name` alone anyway. The first version had the right idea with a row that shares `first_name` but not `last_name`. Can we have both?

```suggestion
                    # update; insert despite matching first_name; insert
                    "first_name": ["B", "B", "X"],
                    "last_name": ["E", "Z", "Y"],
                    "department_name": ["ZZ1", "ZZ2", "ZZ3"],
                    "salary": [999.0, 888.0, 777.0],
```

With that data, joining on only `first_name` fails on DuckDB, Postgres, and Oracle. On Postgres and Oracle it's a "MERGE can't update the same row twice" error rather than an assertion, but it catches it either way.

File: ibis/backends/tests/test_client.py [hunk 6]
```diff
@@ -681,23 +681,64 @@ def _emp(a, b, c, d):
...
+        expected.sort_values(on_cols).reset_index(drop=True),
     )
 
 
+@pytest.mark.notimpl(["polars"], reason="`upsert` method not implemented")
+def test_upsert_empty_on_raises(con, employee_data_1_temp_table, test_employee_data_3):
+    # this validation happens before any backend-specific SQL is built or
+    # executed, so it should raise identically on every backend that
+    # implements `upsert` at all, regardless of that backend's actual MERGE
+    # support
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py]: I'd drop this comment. It also isn't quite accurate right now, since the memtable is registered (which runs SQL) before `on` is validated. If the check moves up into `upsert()` as suggested above, the test is clear enough without it.

**Review by deepyaman — CHANGES_REQUESTED** [review]
Thanks for the update! The implementation looks right to me. Single-column `on` generates exactly the same `MERGE` as before, and I ran the upsert tests against DuckDB, Postgres, and Oracle (plus the no-`MERGE` backends for the empty-`on` test).

Main things to address are the test data not covering the second key column, and the bare `UPDATE` that gets generated when there's nothing left to update, which silently corrupts data on DuckDB. Details inline.

File: ibis/backends/tests/test_client.py [hunk 7]
```diff
@@ -681,23 +681,77 @@ def _emp(a, b, c, d):
...
     backend.assert_frame_equal(
-        result.sort_values("first_name").reset_index(drop=True),
-        expected.sort_values("first_name").reset_index(drop=True),
+        result.sort_values(on_cols).reset_index(drop=True),
+        expected.sort_values(on_cols).reset_index(drop=True),
     )
 
 
+@NO_MERGE_SUPPORT
+@pytest.mark.notyet(["druid"], raises=NotImplementedError)
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py:733]: This is what's failing the Flink jobs in CI. Flink can't create a non-temporary table from in-memory data. The other upsert tests avoid Flink either via `_create_temp_table_with_schema` (which xfails Flink, Druid, and Athena) or with explicit marks like `test_upsert_from_memtable`, so I'd mirror those here:

```suggestion
@pytest.mark.notyet(["druid"], raises=NotImplementedError)
@pytest.mark.notyet(
    ["flink"],
    raises=com.IbisError,
    reason="can't create non-temporary tables from in-memory data",
)
@pytest.mark.notyet(
    ["athena"],
    raises=PyAthenaOperationalError,
    reason="Modifying Hive table rows is only supported for transactional tables",
)
```

Athena isn't in CI, but I'd expect it to fail the same way `test_upsert_from_memtable` does.

File: ibis/backends/tests/test_client.py [hunk 8]
```diff
@@ -681,23 +681,77 @@ def _emp(a, b, c, d):
...
+def test_upsert_on_all_columns(con, temp_table):
+    # when every column is part of `on`, there's nothing left to update; the
+    # `WHEN MATCHED` clause must be omitted rather than emitted with an empty
+    # `SET`, since some backends treat a bare `UPDATE` as "update every
+    # column by position", silently corrupting the row if the source and
+    # target column order differ
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py]: The comment in `_build_upsert_from_table` already covers the why, so I'd drop this one.

```suggestion
```

File: ibis/backends/sql/__init__.py [hunk 9]
```diff
@@ -612,22 +616,32 @@ def _build_upsert_from_table(
...
-                ),
-            ),
+
+        whens = []
+        if non_key_columns:
+            # a bare `WHEN MATCHED THEN UPDATE` with no non-key columns to
+            # set is dangerous: some backends (e.g., DuckDB) treat it as
+            # "update every column by position", which silently corrupts
+            # data whenever the source and target column orders differ.
+            # Omit the clause entirely when there's nothing to update.
```
Comment (deepyaman) [inline ibis/backends/sql/__init__.py]: Nit: can we shorten this a bit?

```suggestion
            # a bare `UPDATE` means a positional `UPDATE SET *` on some
            # backends (e.g., DuckDB), so skip the clause when there's nothing to set
```

File: ibis/backends/tests/test_client.py [hunk 10]
```diff
@@ -751,7 +805,10 @@ def test_upsert_from_expr(
     [
         ({"x": "int64", "y": "float64", "z": "string"}, contextlib.nullcontext()),
         ({"z": "!string", "y": "float32", "x": "int8"}, contextlib.nullcontext()),
-        ({"x": "int64"}, pytest.raises(Exception)),  # No cols to insert
+        (
+            {"x": "int64"},
+            contextlib.nullcontext(),
+        ),  # only the `on` column; no-op update
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py]: Nit: with a shorter comment this fits on one line like the others.

```suggestion
        ({"x": "int64"}, contextlib.nullcontext()),  # only the `on` column
```

File: ibis/backends/tests/test_client.py [hunk 11]
```diff
@@ -783,9 +840,13 @@ def test_upsert_from_memtable(backend, con, temp_table, sch, expectation):
             .reset_index()[list(t1.columns) + [c for c in t2.columns if c not in t1]]
         )
         assert len(result) == len(expected)
+        # normalize null representation before comparing: a partial-column
+        # upsert leaves the omitted columns unset for any newly-inserted
+        # row, and different backends surface that as either `None` or
+        # `NaN`, which pandas will soon treat as non-matching
```
Comment (deepyaman) [inline ibis/backends/tests/test_client.py]: Nit:

```suggestion
        # backends disagree on `None` vs. `NaN` for columns missing from the source
```

**Review by deepyaman — APPROVED** [review]
Thanks, this looks great! I reran the upsert tests on DuckDB and Oracle, and confirmed both of the new/updated tests fail without the fix.

The only thing that needs to change is the Flink/Athena marks on `test_upsert_on_all_columns`, since that's what's breaking CI. The rest are optional nits to trim comments. Approving assuming CI is green after that.
